### DISCUSSION TITLE
Obtaining up-to-date values

### DISCUSSION THREAD

**Discussion by ericchernuka** [mswjs/data#167] [Q&A] [answered]
Is there a best practice to update multiple values and get back the latest values from a new subsequent query? I have two models shown below where when a draft gets updated I need to update the `segment#updatedAt` and `segmentRevision#updatedAt`, but when I retrieve `db.segment.getAll()` the `draftSegment#updatedAt` is the previous value. Does anyone have a suggestion on how I update both models, but ensure the data is up to date across endpoints?

```
  segment: {
    id: primaryKey(faker.datatype.uuid),
    draftSegmentRevision: oneOf('segmentRevision'),
    publishedSegmentRevision: nullable(oneOf('segmentRevision')),
  },
  segmentRevision: {
    id: primaryKey(faker.datatype.uuid),
    updatedAt: () => new Date(),
  },
```

**Comment by ericchernuka** [comment]
I got this working by doing the following:
```js
const now = new Date()

let segment = db.segment.findFirst({ where: { id: { equals: id } })
const draftSegmentRevision = db.segment.update({
  where: { id: { equals: segment.draftSegmentRevision.id } },
  data: { updatedAt: now }
})

segment = db.segment.update({
  where: { id: { equals: id } },
  data: { name: req.body.name, updatedAt: now, draftSegmentRevision }
})
```

This seems to update the references. Just wondering if there's an easier way to handle this.

**Comment by kettanaito** [comment, marked as answer]
Hey, @ericchernuka.

There are a couple of things I can advise based on your implementation.

```js
db.segment.update({
  // No need to query the segment, you can provide the predicate
  // (the segment's "id") to the "update" call directly.
  where: { id: { equals: id } },
  data: {
    updatedAt: now,
    // You can update the relational property from the parent.
    draftSegmentRevision: {
      updatedAt: now,
    }
  }
})
```

Please let me know if the example above does what you intend to do. Thanks. 

## Edit: Collocated update

I came to realize that you can do a collocated update of both `segment` and `segmentRevision` in a single operation. 

```js
db.segment.update({
  where: { id: { equals: id } },
  data: {
    updatedAt: now,
    // You can update properties using a getter function.
    // The getter function is provided with the current value of the field
    // and the current value of the parent entity (segment).
    draftSegmentRevision(_, segment) {
      // Perform the full update operation on the segment revision.
      // Return the result of the update as the next value
      // of the "draftSegmentRevision" property.
      return db.segmentRevision.update({
        where: { id: { equals: segment.draftSegmentRevision.id } },
        data: {
          updatedAt: now
        }
      })
    }
  }
})
```

Collocating these two updates under a single operation is what I'd recommend. Logically, these two updates are connected, so you should connect them operation-wise.

**Reply by ericchernuka** [reply]
Oh interesting. So I don't need to know the draftSegmentationRevision id ahead of time? Ok that's phenomenal.

**Reply by kettanaito** [reply]
You don't need it because you have one-to-one relation between `segment` and `segmentRevision`. This means that 1 segment may have only 1 segment revision associated with it. You should be able to update it directly in that case.

**Reply by ericchernuka** [reply]
Amazing. My mind continues to be blown with the attention to detail of these APIs.

**Reply by kettanaito** [reply]
Querying the data _after_ the update must always return you the next (latest) state of the data. Entities are updated directly, while relationships are updated by reference (only the actual referenced entity is updated, all pointers to it never store any intermittent value to become potentially out-of-date). 

If that doesn't happen, that's likely an issue with the update or with the library. Please, if you encounter any strange behavior, don't hesitate to comment here. Including a reproduction repo would be ideal in that case.

**Reply by ericchernuka** [reply]
@kettanaito, Just implementing this now and it looks like the TypeScript types don't like the omission of the `id` of the `draftSegmentRevision`.

![image](https://user-images.githubusercontent.com/243155/143491320-7fde5928-33b0-4073-9c4d-b23f84d9b738.png)

Console error.
```
Invariant Violation: Failed to define a relational property "draftSegmentRevision" on "segment": referenced entity "undefined" ("id") does not exist.
```

**Reply by kettanaito** [reply]
Nice, I think it's an issue. May I please ask you to create one in this repo?

**Reply by ericchernuka** [reply]
You bet.

**Reply by kettanaito** [reply]
A follow-up question: the library throws an exception when _defining_ the "draftSegmentRevision" relational property. Looks like it doesn't exist at the time you update the segment. Could you please include how you create the segment and define the revision on it?

**Reply by ericchernuka** [reply]
I have a seed file that builds all the resources when the endpoints mount. This update is called after I navigate into a segment and attempt to update it so its there because the data I'm seeing in my UI is from the draftSegmentRevision.

Here's an example:
```typescript
  const segmentBaseTemplates: Partial<Segment>[] = [
    {
      id: 'f7c419d9-b0d2-5be2-b7c2-faa3b4777edb',
      name: 'LATAM medium',
    }
  ];

  segmentBaseTemplates.forEach((segmentTemplate) => {
    const draftSegmentRevision = dbInstance.segmentRevision.create();

    dbInstance.segment.create({
      id: segmentTemplate.id,
      name: segmentTemplate.name,
      draftSegmentRevision,
    });
  });
```

The update logic is then in an endpoint and pretty much matches the updater you suggested above.

**Comment by kettanaito** [comment]
As we are solving #168, you can update the segment and its revision like this:

```js
// Update the segment.
const segment = db.segment.update({
  where: { id: { equals: id } },
  data: { updatedAt: now }
})

// Update the revision associated with this segment.
db.segmentRevision.update({
  where: { id: { equals: segment.draftSegmentRevision.id } },
  data: { updatedAt: now }
})
```

**Reply by ericchernuka** [reply]
Sadly that doesn't work either. The `segmentRevision` is indeed updated, but when referred through the `segment` it is still the previous value. Only way I've been able to make this work is by passing the return from the updated `segmentRevision` into the subsequent update to the `segment`

**Reply by kettanaito** [reply]
Could you please double-check that you're using the latest version of `@mswjs/data`? That issue you're describing sounds like an issue we've addressed recently.

**Reply by ericchernuka** [reply]
Looks like I'm running `0.8.0`
```
"@mswjs/data@^0.8.0":
  version "0.8.0"
  resolved "https://registry.yarnpkg.com/@mswjs/data/-/data-0.8.0.tgz#6d1598d5542774a7c41a6a3da27ecb4a073b07c7"
```

**Reply by kettanaito** [reply]
Interesting. It appears that the relational property gets resolved (loses its `[getter]`) on the parent entity when it's returned from the `update` method. That's an issue!

**Reply by kettanaito** [reply]
The issue is described in #170 and addressed in https://github.com/mswjs/data/pull/172. Thank you for reporting that!

**Reply by kettanaito** [reply]
My suggestion above works as expected in [0.8.1](https://github.com/mswjs/data/releases/tag/v0.8.1). @ericchernuka, please update and let me know.

**Reply by ericchernuka** [reply]
You're the best! Works as intended. Still looking for time to look into the other issue, but this is great for now.

**Reply by kettanaito** [reply]
Please, if you'd like to support my effort, consider [sponsoring our project](https://opencollective.com/mswjs) on Open Collective. I love helping people build better products through quality testing. Doing that full-time would be my dream. Each contribution brings me closer to that dream. Thank you!
